keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,396 @@
+        klog.Infof("synchronizing PersistentVolume[%s]: the volume was requested to be deleted but still" +
+            " associated with the claim %s", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
+        _, err := ctrl.updateRequestToDeleteVolumeAnnotations(volume)
+        if err != nil {
+            klog.V(4).Infof("synchronizing PersistentVolume[%s]: failed to add the delete volume annotation", volume.Name)
+        }
+        klog.Infof("successfully added annotation to indicate the volume is to be deleted.")
+    }
+}
+
+// Remaining code below skipped for brevity...
+// Please refer kubernetes/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go for the full code.
+```
+
+When `deleteVolumeOperation` executes, `NotFound` errors while retrieving the PV is ignored if the annotation is found.
+
+```go
+// deleteVolumeOperation deletes a volume. This method is running in standalone
+// goroutine and already has all necessary locks.
+func (ctrl *PersistentVolumeController) deleteVolumeOperation(volume *v1.PersistentVolume) (string, error) {
+	klog.V(4).Infof("deleteVolumeOperation [%s] started", volume.Name)
+	var requestToDeleteAnn bool
+	// This method may have been waiting for a volume lock for some time.
+	// Previous deleteVolumeOperation might just have saved an updated version, so
+	// read current volume state now.
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil { //<===============================Changed here START
+		if !apierrors.IsNotFound(err) {
+			return "", nil
+		} else {
+			// Check if the volume is annotated with explicit request to delete.
+			// In this case due to race with the pv protection controller, it was not possible to retrieve the latest PV
+			// state. However, the annotation clearly indicates the volume deletion can proceed.
+			requestToDeleteAnn = metav1.HasAnnotation(volume.ObjectMeta, pvutil.AnnRequestToDeleteVolume)
+			if !requestToDeleteAnn {
+				klog.Infof("no request to delete annotation found on the volume %s", volume.Name)
+				// Does not have the annotation, throw the NotFound error.
+				return "", err
+			} else {
+				klog.Infof("request to delete annotation found on the volume %s", volume.Name)
+				newVolume = volume
+			}
+		}
+	} //<===============================Changed here END
+    // Note: no need to check for deletion timestamp, the schedule operation ensures that duplicate delete
+    // are not processed.
+    if newVolume.GetDeletionTimestamp() != nil {//<===========================Check to be removed
+        klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+        return "", nil
+    }
+// Remaining code below skipped for brevity...
+// Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+
+The above changes ensure that the plugin receives the volume deletion request.
```

Comment:
The above changes The above changes ensure that the plugin receives the volume deletion request only once. If deletion fails, it will not be retried, leaking the volume again. In addition, if kube-controller-manager is restarted around these parts, any knowledge about the PV is lost, as it may be already deleted in the API server when new kube-controller-manager starts.

Reply:
@jsafrane Isn't the volume added to a queue for retries if deletion fails?

Reply:
I've changed the approach to use finalizer instead of annotations, this would prevent the pv from deleting the API server.